Core/SubterfugeCore/Core/Timing/TimeMachine.cs

```diff
@@ -1,166 +1,258 @@
-        }
-
-        /// <summary>
-        /// For debugging. Rewinds the timeMachine by a specified number of ticks.
-        /// </summary>
-        /// <param name="ticks">The number of ticks to rewind</param>
-        public void Rewind(int ticks)
-        {
-            GameTick tick = this.GetCurrentTick().Rewind(ticks);
-            this.GoTo(tick);
-        }
-
-        /// <summary>
-        /// Gets a list of the queued events
-        /// </summary>
-        /// <returns>A list of the events in the future event queue</returns>
-        public List<GameEvent> GetQueuedEvents()
-        {
-            List<GameEvent> gameEvents = new List<GameEvent>();
-            foreach(GameEvent gameEvent in this._futureEventQueue.GetQueue()){
-                if (gameEvent != null){
-                    gameEvents.Add(gameEvent);
-                }
-            }
-            return gameEvents;
-        }
-    }
+	public class TimeMachine
+	{
+		// List of known events
+		private ReversePriorityQueue<GameEvent> _pastEventQueue = new ReversePriorityQueue<GameEvent>();
+		private PriorityQueue<GameEvent> _futureEventQueue = new PriorityQueue<GameEvent>();
+
+		private VisionManager _visionManager;
+
+		// Current representation of the game state
+		private GameState _gameState;
+
+		/// <summary>
+		/// Creates a new instance of the TimeMachine. You will likely never need to call this as this is created in the
+		/// `Game` object when the game is created.
+		/// </summary>
+		/// <param name="state">The initial GameState</param>
+		public TimeMachine(GameState state)
+		{
+			_gameState = state;
+			_visionManager = new VisionManager(state.GetPlayers());
+		}
+
+		/// <summary>
+		/// Get the time machine's current state
+		/// </summary>
+		/// <returns>The GameState at the current time of the TimeMachine</returns>
+		public GameState GetState()
+		{
+			return this._gameState;
+		}
+
+		public VisionManager GetVisionManager()
+		{
+			return _visionManager;
+		}
+
+		/// <summary>
+		/// Adds an event to the future event queue
+		/// </summary>
+		/// <param name="gameEvent">The game event to add to the Queue</param>
+		public void AddEvent(GameEvent gameEvent)
+		{
+			this._futureEventQueue.Enqueue(gameEvent);
+		}
+
+		/// <summary>
+		/// Removes a GameEvent from the game.
+		/// </summary>
+		/// <param name="gameEvent">The GameEvent to remove from the queue</param>
+		public void RemoveEvent(GameEvent gameEvent)
+		{
+			if (this._futureEventQueue.GetQueue().Contains(gameEvent))
+			{
+				this._futureEventQueue.Remove(gameEvent);
+			}
+			else
+			{
+				// Go to 1 tick before the event occurs.
+				GameTick currentTick = GetCurrentTick();
+				GoTo(gameEvent);
+				Rewind(1);
+				this._futureEventQueue.Remove(gameEvent);
+				GoTo(currentTick);
+			}
+		}
+
+		/// <summary>
+		/// Advances the game state to the next tick. Creates (or recalculates) vision events for the tick.
+		/// </summary>
+		public void NextTick()
```

Comment:
I'm curious to see how fast this works and if it is fast enough such that a player can constantly scroll into the future uninterrupted